src/io.c

```diff
@@ -1437,20 +1437,20 @@ _dispatch_fd_entry_create_with_fd(dispatch_fd_t fd, uintptr_t hash)
 				int result = ioctlsocket((SOCKET)fd, (long)FIONBIO, &value);
 				(void)dispatch_assume_zero(result);
 			} else {
-				// Try to make writing nonblocking, although pipes not coming
-				// from Foundation.Pipe may not have FILE_WRITE_ATTRIBUTES.
+				// The _dispatch_pipe_monitor_thread expects pipes to be
+				// PIPE_WAIT and exploits this assumption by using a blocking
+				// 0-byte read as a synchronization mechanism.
```

Comment:
`_dispatch_pipe_monitor_thread` is the reader and this code is for the writer. We shouldn't assume that the reader thread is using libdispatch. We could just assert that it's not NOWAIT and comment that the writer is misbehaved if NOWAIT is set, regardless of what the reader does.

Reply:
Hmm, I'm not following. `_dispatch_pipe_monitor_thread` isn't actually the reader, it just waits until data is available on the pipe and then signals to the _real_ reader that data is available. I don't think you can get around using `_dispatch_pipe_monitor_thread` if you are writing data through a Win32 pipe on libdispatch, so I think the comment is relevant here.